File: r/number_of_flights_change.R
networks_popular_flights <- function(){
  
  library(DBI)
  library(igraph)
  library(dplyr)
  
  conn <- dbConnect(RSQLite::SQLite(), dbname = '/Users/levbarbash/LSE/LondonProg/Coursework/airline2.db')
  
  origin_dest <- list()
  for (year in 2006:2007) {
    query <- glue::glue("
      SELECT Origin AS Origin, Dest AS Dest
      FROM ontime
      WHERE ontime.Cancelled = 0 AND ontime.Diverted = 0 AND ontime.Year = {year}
    ")
    temp_sers <- dbGetQuery(conn, as.character(query))
    
    temp_sers <- temp_sers %>%
      group_by(Origin, Dest) %>%
      summarise(Weight = n()) %>%
      arrange(desc(Weight))
    
    origin_dest[[year - 2006 + 1]] <- temp_sers
  }
  dbDisconnect(conn)
  
  # Merge dataframes
  origin_dest_merged <- inner_join(origin_dest[[1]] %>% head(100), origin_dest[[2]], by = c("Origin", "Dest"))
  origin_dest_merged <- origin_dest_merged %>% 
    rename("Weght2006" = "Weight.x",
           "Weght2007" = "Weight.y")
  
  # Create network graphs
  graphs <- list(graph_from_data_frame(select(origin_dest_merged, Origin, Dest), directed = FALSE), graph_from_data_frame(select(origin_dest_merged, Origin, Dest), directed = FALSE))
  min_weight <- min(origin_dest_merged[,c(3,4)])
  max_weight <- max(origin_dest_merged[,c(3,4)])
  
  for (i in 1:2) {
    g <- graphs[[i]]
    year <- 2006 + i - 1
    weights <- origin_dest_merged[, i+2]
    weightsm <- lapply(weights, function(x) (x - min_weight) / (max_weight - min_weight))
    plot_title <- paste("100 most popular routes in", year)
    
    plot(g, layout = layout.circle(g), edge.color = "blue", edge.width = unlist(weightsm), main = plot_title, vertex.size=0.1)
  }
}

File: python/number_of_flights_change.py
def networks_popular_flights():
    
    import sqlite3
    import sql_query
    import pandas as pd
    import numpy as np
    import seaborn as sns
    import matplotlib.pyplot as plt
    import networkx as nx
    
    origin_dest = []
    for year in range(2006, 2008):
        temp_sers = sql_query.execute_sql_query('''
        SELECT Origin AS Origin, Dest as Dest
        FROM ontime
        WHERE ontime.Cancelled = 0 AND ontime.Diverted = 0 AND ontime.Year = ''' + str(year) + '''
        ''')
        temp_sers = temp_sers.rename(columns={0:'Origin', 1:'Dest'})
        temp_sers = temp_sers.groupby(['Origin', 'Dest']).size().reset_index().rename(columns={0:'Weight' + str(year)})
        temp_sers = temp_sers.sort_values(by='Weight' + str(year), ascending=False)
        origin_dest.append(temp_sers)
    origin_dest = pd.merge(origin_dest[0].head(100), origin_dest[1], on=['Origin', 'Dest'])
    
    graphs = [nx.Graph(), nx.Graph()]
    min_weight = origin_dest[['Weight2006', 'Weight2007']].min().min()
    max_weight = origin_dest[['Weight2006', 'Weight2007']].max().max()
    plt.subplots(figsize=(25,10))

    for g, year, sbplt in zip(graphs, [2006, 2007], [121, 122]):
        for index, row in origin_dest.iterrows():
            g.add_edge(u_of_edge=row['Origin'], v_of_edge=row['Dest'], weight=row['Weight' + str(year)])
        edges = g.edges()
        weights = [g[u][v]['weight'] for u,v in edges]
        weightsm = list(map(lambda x: (x - min_weight) / (max_weight - min_weight), weights))
        plt.subplot(sbplt)
        pos=nx.circular_layout(g)
        plt.title('100 most popular routes in ' + str(year))
        nx.draw_networkx(g, pos, node_color='blue', edge_color='blue', width=weightsm, node_size=1)
    